### ISSUE TITLE
`navi-scripts build` fails with react hooks

### ISSUE THREAD

**Issue by joeporpeglia** [frontarm/navi#41]
When running `navi-scripts build` using hooks it fails with the following error:

**Overview**
```
[ohshit] An error occured while building your app
Minified React error #298; visit https://reactjs.org/docs/error-decoder.html?invariant=298 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
Invariant Violation: Minified React error #298; visit https://reactjs.org/docs/error-decoder.html?invariant=298 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
```

[The unminified error message](https://reactjs.org/docs/error-decoder.html?invariant=298) states:

> Hooks can only be called inside the body of a function component.

I know hooks are still in alpha but it might be worth investigating this issue before they release.

**Steps to reproduce**
1. Clone the navi repo
2. `cd` into `navi/examples/create-react-blog` 
3. Update react and react-scripts `16.7.0-alpha.2`
4. Replace the App component in `App.js` to use a hook (see below)
5. Run `yarn build` for the example

**Hook usage in App.js**
```
export function App(props) {
  const [state] = useState(true)
  return (
    <Nav.Provider navigation={props.navigation}>
      <Nav.Loading>
        {isLoading => (
          <>
            <AppBusyIndicator show={isLoading} />
            <nav>
              <Nav.Link href="/">Home</Nav.Link>
              {state ? 'state is true' : 'state is false'}
            </nav>
            <main>
              <Nav.NotFoundBoundary render={renderNotFound}>
                <Nav.Route />
              </Nav.NotFoundBoundary>
            </main>
          </>
        )}
      </Nav.Loading>
    </Nav.Provider>
  )
}
```

**Comment by joeporpeglia** [comment]
[This comment](https://github.com/facebook/react/issues/14317#issuecomment-441693397) gives me feeling the build command is using two copies of react. Maybe instead of [importing `react` and `react-dom` in the config processor](https://github.com/frontarm/navi/blob/c746765a2d372b85753fa935cdc46ecca959b97f/packages/navi-scripts/lib/config.js#L124-L125) we should [read it from the global object](https://github.com/frontarm/navi/blob/c746765a2d372b85753fa935cdc46ecca959b97f/packages/navi-scripts/lib/createScriptRunner.js#L97) and pass it to `renderPageToString` like we're doing with the global navi object.

**Comment by joeporpeglia** [comment]
I originally ran into this issue yesterday on a fresh project using all the latest dependencies. The example project was just an easier way to describe repro steps. Ill try to get a codesanbox together with the example app updated if that would help.

**Comment by joeporpeglia** [comment]
Most of the solutions I can think of add a lot of coupling between navi and the CRA build process and I really like the layered separation between them right now. Maybe hooks users can pass an async function to `app` that dynamically imports `ReactDOMServer` and returns it? That would at least split it out of the production bundle, but the module mapping would still be included in the production webpack manifest/runtime. I'll keep thinking on it..

**Comment by jamesknelson** [comment]
I've just pushed an update which should fix this, and it will be available in Navi 0.11.

The idea is that `Navi.app()` (now called `register()`) looks for the "react" and "react-navi" modules, and if it exists, it exports them so that the build script can use them. If they don't exist, nothing changes, preserving the decoupled design.

Then, navi-scripts hooks into `require()` and makes it return the instance of "react" and "react-navi" exported by the app bundle. You can override this behavior or add extra shared modules by passing a `sharedModules` object to `register()`.

Will be publishing the docs and releasing this when React 16.8 is published.
